libvast/src/system/index.cpp

```diff
@@ -1253,66 +1253,81 @@ index(index_actor::stateful_pointer<index_state> self,
               .stats = std::move(aps.stats),
             };
             // Update the index statistics. We only need to add the events of
             // the new partition here, the subtraction of the old events is
             // done in `erase`.
             for (const auto& [name, stats] : result.stats.layouts)
               self->state.stats.layouts[name].count += stats.count;
+
             if (keep == keep_original_partition::yes) {
-              self
-                ->request(self->state.meta_index, caf::infinite, atom::merge_v,
-                          new_partition_id, aps.synopsis)
-                .then(
-                  [self, rp, new_partition_id, result](atom::ok) mutable {
-                    self->state.persisted_partitions.insert(new_partition_id);
-                    rp.deliver(result);
-                  },
-                  [rp](const caf::error& e) mutable {
-                    rp.deliver(e);
-                  });
+              if (aps.synopsis)
+                self
+                  ->request(self->state.meta_index, caf::infinite,
+                            atom::merge_v, new_partition_id, aps.synopsis)
+                  .then(
+                    [self, rp, new_partition_id, result](atom::ok) mutable {
+                      self->state.persisted_partitions.insert(new_partition_id);
+                      rp.deliver(result);
+                    },
+                    [rp](const caf::error& e) mutable {
+                      rp.deliver(e);
+                    });
+              else
+                rp.deliver(result);
             } else {
               // Pick one partition id at random to be "transformed", all the
-              // other ones are "deleted" from the meta index.
-              auto old_partition_id = old_partition_ids.at(0);
-              self
-                ->request(self->state.meta_index, caf::infinite,
-                          atom::replace_v, old_partition_id, new_partition_id,
-                          aps.synopsis)
-                .then(
-                  [self, rp, old_partition_id, new_partition_id,
-                   result](atom::ok) mutable {
-                    self->state.persisted_partitions.insert(new_partition_id);
-                    self
-                      ->request(static_cast<index_actor>(self), caf::infinite,
-                                atom::erase_v, old_partition_id)
-                      .then(
-                        [=](atom::done) mutable {
-                          rp.deliver(result);
-                        },
-                        [=](const caf::error& e) mutable {
-                          rp.deliver(e);
+              // other ones are "deleted" from the meta index. If the new
+              // partition is empty, all partitions are deleted.
+              auto start = size_t{0};
+              if (aps.synopsis) {
+                start = 1;
+                auto old_partition_id = old_partition_ids.at(0);
```

Comment:
You can avoid the explicit index management by choosing the last partition with `back()` and `pop_back()` instead.